### PR TITLE
chore: Set up unit testing, eslint, and TypeScript configuration for design-system-react

### PR DESCRIPTION
**Pull request by georgewrmarshall**
## **Description**

This pull request sets up the `design-system-react` package with proper unit testing, eslint, and TypeScript configurations. Key changes include:

- Integration of testing dependencies, including `@testing-library/jest-dom` and `@testing-library/react`.
- Configuration of eslint rules to support TypeScript files
- Alignment of TypeScript settings to ensure compatibility with eslint and testing configurations.
- Addition of initial test examples to validate the setup and document testing standards.

## **Related issues**

Fixes: [MetaMask/metamask-design-system#77](https://github.com/MetaMask/metamask-design-system/issues/77)

## **Manual testing steps**

1. Run `yarn install` to install the updated dependencies.
2. Verify that eslint runs successfully without errors in TypeScript and React files.
3. Run `yarn workspace @metamask/design-system-react test` to confirm all initial tests pass.
4. Check for correct TypeScript compilation and ensure no type errors are thrown.

## **Pre-merge author checklist**

- [x] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs)
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g., pulled and built the branch, run the app, and verified the code changes).
- [ ] I confirm that this PR addresses all acceptance criteria described in the issue it closes and includes necessary testing evidence such as recordings and/or screenshots.

### CODE REVIEW

**Comment by georgewrmarshall** [comment]
@metamaskbot publish-preview

File: .eslintrc.js [hunk 1]
```diff
@@ -13,8 +13,11 @@ module.exports = {
   ],
   overrides: [
     {
-      files: ['*.test.{ts,js}'],
+      files: ['*.test.{ts,tsx,js}'],
       extends: ['@metamask/eslint-config-jest'],
+      rules: {
+        '@typescript-eslint/no-shadow': ['error', { allow: ['Text'] }],
```
Comment (georgewrmarshall) [inline .eslintrc.js:19]: Prevents this error

Screenshot 2024-10-31 at 2 59 17 PM

`Text` is frequently used as a component name, especially in UI libraries like React Native or styled-components. Allowing Text helps prevent unnecessary linting errors when importing or defining UI components.

File: jest.config.packages.js [hunk 2]
```diff
@@ -22,7 +22,7 @@ module.exports = {
   collectCoverage: true,
 
   // An array of glob patterns indicating a set of files for which coverage information should be collected
-  collectCoverageFrom: ['./src/**/*.ts'],
+  collectCoverageFrom: ['./src/**/*.{ts,tsx}'],
```
Comment (georgewrmarshall) [inline jest.config.packages.js:25]: Updating global jest configurations for packages to collect testing coverage from `tsx` files

File: packages/design-system-react/jest.config.js [hunk 3]
```diff
@@ -23,4 +23,10 @@ module.exports = merge(baseConfig, {
       statements: 100,
     },
   },
+
+  // Add the following to include the Jest setup file
+  setupFilesAfterEnv: ['<rootDir>/jest.setup.ts'],
+
+  // **Add this line to set the test environment to jsdom**
```
Comment (georgewrmarshall) [inline packages/design-system-react/jest.config.js:30]: Improve these comments

File: packages/design-system-react/jest.setup.ts [hunk 4]
```diff
@@ -0,0 +1,2 @@
+// eslint-disable-next-line import/no-unassigned-import
+import '@testing-library/jest-dom';
```
Comment (georgewrmarshall) [inline packages/design-system-react/jest.setup.ts:2]: Disables this linting rule `error  Imported module should be assigned`

Screenshot 2024-10-31 at 3 04 49 PM

File: packages/design-system-react/package.json [hunk 5]
```diff
@@ -51,6 +51,8 @@
     "@metamask/design-system-tailwind-preset": "workspace:^",
     "@storybook/react": "^8.3.5",
     "@storybook/test": "^8.3.5",
+    "@testing-library/jest-dom": "^6.6.3",
+    "@testing-library/react": "^16.0.1",
```
Comment (georgewrmarshall) [inline packages/design-system-react/package.json:55]: Adding testing library dev dependencies for testing library

File: packages/design-system-react/src/Button.test.tsx [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+    render(<Button onClick={handleClick}>Click</Button>);
+    fireEvent.click(screen.getByText('Click'));
+    expect(handleClick).toHaveBeenCalledTimes(1);
+  });
+
+  it('passes additional props to the button element', () => {
+    render(<Button aria-label="custom-button">Accessible Button</Button>);
+    expect(screen.getByLabelText('custom-button')).toBeInTheDocument();
+  });
+});
```
Comment (georgewrmarshall) [inline packages/design-system-react/src/Button.test.tsx:29]: Adding example tests for button is temporary

File: packages/design-system-react/src/Text.test.tsx [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+    const { container } = render(
+      <Text variant={TextVariant.BodySm} className="custom-class">
+        Custom Class
+      </Text>,
+    );
+    expect(container.firstChild).toHaveClass(
+      'text-s-body-sm font-s-body-sm leading-s-body-sm tracking-s-body-sm lg:text-l-body-sm lg:font-l-body-sm lg:leading-l-body-sm lg:tracking-l-body-sm custom-class',
+    );
+  });
+});
```
Comment (georgewrmarshall) [inline packages/design-system-react/src/Text.test.tsx:30]: Adding example tests for Text is temporary

File: packages/design-system-react/src/index.test.ts [hunk 8]
```diff
@@ -1,9 +1,11 @@
...
-    const name = 'Huey';
-    const result = greeter(name);
-    expect(result).toBe('Hello, Huey!');
+describe('Design System React Index', () => {
+  it('should export Button component', () => {
+    expect(Button).toBeDefined();
+  });
+
+  it('should export Text component', () => {
+    expect(Text).toBeDefined();
```
Comment (georgewrmarshall) [inline packages/design-system-react/src/index.test.ts:9]: Updating index test to reflect index.ts

File: packages/design-system-react/src/index.ts [hunk 9]
```diff
@@ -1,9 +1,2 @@
...
- * Example function that returns a greeting for the given name.
- *
- * @param name - The name to greet.
- * @returns The greeting.
- */
-export default function greeter(name: string): string {
-  return `Hello, ${name}!`;
-}
+export { Button } from './Button';
+export { Text } from './Text';
```
Comment (georgewrmarshall) [inline packages/design-system-react/src/index.ts:2]: Exporting Button and Text from index so we can test package locally

File: packages/design-system-react/tsconfig.build.json [hunk 10]
```diff
@@ -4,7 +4,13 @@
...
     "outDir": "./dist",
     "rootDir": "./src",
-    "jsx": "react"
+    "jsx": "react",
+    "types": [
+      "react",
+      "jest",
+      "@testing-library/react",
+      "@testing-library/jest-dom"
+    ]
```
Comment (georgewrmarshall) [inline packages/design-system-react/tsconfig.build.json:13]: - `@testing-library/react`: Provides TypeScript type definitions for React Testing Library. This allows TypeScript to understand the types used in our tests, enabling features like autocomplete and type checking.
- `@testing-library/jest-dom` : Offers custom Jest matchers that are tailored for DOM node assertions. Including its types ensures that these matchers are recognized and type-checked by TypeScript.

Without we get this build error running `yarn build`

Screenshot 2024-10-31 at 3 07 47 PM

**Review by brianacnguyen — APPROVED** [review]

**Pull request merged**